### ISSUE TITLE
Playback Quality Select Feature

### ISSUE THREAD

**Issue by Liqianyu** [CastagnaIT/plugin.video.netflix#1265]
## Feature request
I'm submitting a proposal to ...
  - [X] new feature
  - [ ] change an existing feature

### Describe your request

I wish the Netflix add-on would add [playback quality selection](https://www.matthuisman.nz/2019/09/new-quality-select-in-my-add-ons.html) similar to that developed by MATT HUISMAN. Why do you want this? Because inputstream adaptive has never been able to control based on resolution. And now it is constantly adapting the video stream. This is a good idea but it makes Kodi constantly flicker the screen (matching resolution and frame rate).

#### Describe the use case

If  have this feature, can avoid the ISA adaptive streaming and force it at the best definition. Even without the adaptive streaming issue, previously (Kodi 19) I had to manually adjust to the highest definition at the beginning of each session. (ISA set to auto does not guarantee at best definition)

#### Describe an alternative

If ISA can solve the adaptive problems that still exist with current manual streams. Or even introduce playback quality control features. Then there is no need for additional processing by add-ons.

#### Development tips

Maybe refer to matthuisman's implementation?

#### Specific details

https://forum.kodi.tv/showthread.php?tid=364745

**Comment by CastagnaIT** [comment]
I start by saying that you are requesting a feature for the next Kodi 20, which at present it is at the first stage of development, so in the ALPHA state.

This means that the current Kodi/ISA development can change from one month to the next, and invest time in creating that will shortly become obsolete or outdated doesn't make sense.

> Because inputstream adaptive has never been able to control based on resolution.

Until to Kodi 19 by default ISA always select the higher resolution to match display resolution, if this not happens means that you have changed some settings in wrong way, or you have a lower display resolution then is needed to enable `Ignore display resolution`.
ISA allow also to limit the max resolution, and allow also customizations per bandwidth in his settings, then what you say is not so right

> And now it is constantly adapting the video stream. This is a good idea but it makes Kodi constantly flicker the screen (matching resolution and frame rate).

Always on Kodi 20, at current stage is still in development and devs already know the problems,
it has already been planned to add a way to disable stream adaptive switching feature, by addon/settings, until it becomes stable enough

I don't really like the manual stream selection GUI dialog as mentioned in "matthiusman" addons,
because this should not be the task of the addon and neither the user should have to worry about it, Kodi/ISA should manage it in trasparent way without interactions.

Basically this "feature" force the removal of streams submitted via ISA proxy and this thing must take into account many variables also with multiple codecs with same resolutions and or different refresh rate, so it should also manage priorities of codec types.

Currently this is not something i will consider to study/implementing
Maybe we'll talk about it when Kodi 20 will be in RC/Final version and in the worst case scenario the situation will not have improved

**Comment by Liqianyu** [comment]
Thanks for the quick response, I think you have some valid points. So for now I'm avoiding the adaptive bitrate issue by using Kodi 19.
I know ISA can make it force to work at the highest bitrate by setting Min. But can it make the audio stream work at the optimal bitrate? I'm not sure... It seems to work? I think this method is a bit unclear, since different streaming services offer different bitrates.
I think matthuisman's playback quality selection offers a lot more freedom. It could solve the glitches that Kodi 20 adaptive bitstreaming currently causes. But as you say, this is probably not a feature that should be implemented by add-ons. Should it be done by ISA?

**Comment by jakermx** [comment]
On ISA Settings, Set to Manually Select Streams.... and there you go,  the MAximum Resolution Available will be shown based on your Subscripption,  Widevine Device Level and Available Codecs, on NF AddOn there is a Max Resolution Limit, that allows LowEnd Devices to  avoid ISA Configuration, but if you set it to 4K,  you can select streams on ISA if set to Manual.

**Comment by Liqianyu** [comment]
@jakermx 
Min.Bandwidth does not guarantee that ISA will choose the best bitrate. For example, if I set it to 16384000, it should always use the best bitrate for Netflix. However, it doesn't. Once I set ISA to manually select the stream, then every video will start playing at the lowest resolution. I have to manually adjust to the highest resolution, and I don't understand what the resolution limit of the Netflix add-on does.

**Comment by jakermx** [comment]
I usually think that NF Addon is a Broker, between  NF and Kodi Player (ISA), so...The AddOn, retreives the information from the server and once formated , passes it to the KODI PLayer.... based on it, the Resolution Limit, just adquiere the manifest data  below that limit...

**Comment by jakermx** [comment]
I think, that the most that the AddOn could do for your request is....showing an extra Dialog when click on play...that let you choose available resolutions of thaat specific title.. so then it passes just that manifest and license data to KODI Player....

**Comment by Liqianyu** [comment]
I tested again the add-on's "Limit video stream resolution to", which serves to limit a maximum resolution. For example, if it is set at HD 720p, ISA will not get a video stream resolution higher than 720p. Consider that this option for UHD 4K does not make sense at the moment, after all, Netflix currently has no content above 4K.
All in all ISA's automatic stream selection has no real control over stream resolution. (I set the minimum bandwidth and it still seems to automatically select a resolution below the minimum bandwidth) and if you do it manually you have to manually select the video stream every time the video is played. That's why the playback quality selection feature developed by matthuisman makes sense. It allows the stream to work at the best bitrate, although in some cases this means buffering. But... At least the user can choose it.
But as CastagnaIT said, I also prefer that the functionality should be integrated in ISA rather than handled by the add-on itself. But until that feature is currently not available in ISA, matthuisman's playback quality select fills that need. And it allows to turn off or bypass that processing.

**Comment by jakermx** [comment]
As I said, content resolution depends on some factors:
-Subscription
-Widevine Devicee Leveling
-Devicee Codec capabilities 
-Bandwidth(Dont Affect Aviability)

I have 4k NF Content available, but not on all my devices....so consider it.... also,  try set defaults on player OSD, when selecting high Resolutions on ISA---

Cheers

**Comment by jakermx** [comment]
You may like to try the following Titles
-Test Patterns
-Moving Art
-Our Planet

**Comment by Liqianyu** [comment]
@jakermx 
Do you mean to use "Set as default for all movies" after manually selecting a stream to save it as default?
Thanks for the reminder, I'll give it a try.

**Comment by jakermx** [comment]
Yup

**Comment by Liqianyu** [comment]
@jakermx 
Have you tried it to work? I manually selected the best bitrate stream after Set as default for all movies, but it didn't work. Everything else will still play in the default first bitrate stream (lowest).
